src/bun.js/net.exports.js

```diff
@@ -53,9 +53,246 @@ export function isIP(s) {
   return 0;
 }
 
+const { Duplex } = import.meta.require("node:stream");
+
+export class Socket extends Duplex {
+  bytesRead = 0;
+  bytesWritten = 0;
+  connecting = false;
+  localAddress = "127.0.0.1";
+  #readQueue = [];
+  remotePort;
+  #socket;
+  timeout = 0;
+  #writeQueue;
+
+  constructor(options) {
+    super({
+      allowHalfOpen: options?.allowHalfOpen || false,
+      readable: true,
+      writable: true,
+    });
+    options?.signal?.once("abort", () => this.destroy());
+    this.once("connect", () => this.emit("ready"));
+    // TODO support `options.fd`
+  }
+
+  address() {
+    return {
+      address: this.localAddress,
+      family: this.localFamily,
+      port: this.localPort,
+    };
+  }
+
+  get bufferSize() {
+    return this.writableLength;
+  }
+
+  connect(port, host, connectListener) {
+    // TODO support IPC sockets
+    if (typeof host == "function") {
+      connectListener = host;
+      host = undefined;
+    }
+    if (typeof port == "object") {
+      var {
+        port,
+        host,
+        // TODOs
+        localAddress,
+        localPort,
+        family,
+        hints,
+        lookup,
+        noDelay,
+        keepAlive,
+        keepAliveInitialDelay,
+      } = port;
+    }
+    const self = this;
+    let ended = false;
+    self.connecting = true;
+    self.remotePort = port;
+    if (connectListener) self.on("connect", connectListener);
+    Bun.connect({
+      hostname: host || "localhost",
+      port: port,
+      socket: {
```

Comment:
Could we move this to be an object created only once at the top of this file? Then, we can assign `data` on `socket` to the instance

```
const self = socket.data;
```

Reply:
That way if creating many connections it doesn't re-create the object every time

Reply:
not sure if you are referring to the `self` a few lines up, but that's a `net.Socket` instance, not of `bun`'s socket...

Reply:
Done in https://github.com/oven-sh/bun/pull/1701/commits/81b7104e6517d030ce8aea87609f4b938509a1c3